File: core/src/main/java/com/views/textview/countdown/LCountDownView.java
package com.views.textview.countdown;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.R;
import com.core.utilities.LAnimationUtil;
import com.daimajia.androidanimations.library.Techniques;
import com.interfaces.CallbackAnimation;
//TODO convert kotlin
public class LCountDownView extends RelativeLayout {
    private final String logTag = getClass().getSimpleName();
    private TextView tvCountDown;

    public LCountDownView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public LCountDownView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_l_count_down, this);
        this.tvCountDown = findViewById(R.id.tv_count_down);
    }

    public void start(int number) {
        this.number = number;
        doPerSec();
    }

    public void setShowOrHide(boolean isShow) {
        if (tvCountDown != null) {
            tvCountDown.setVisibility(isShow ? VISIBLE : INVISIBLE);
        }
    }

    private int number;

    public interface Callback {
        void onTick();

        void onEnd();
    }

    private Callback callback;

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    private void doPerSec() {
        if (tvCountDown != null) {
            tvCountDown.setText(String.valueOf(number));
            LAnimationUtil.Companion.playDuration(tvCountDown, Techniques.FlipInX, 1000, new CallbackAnimation() {
                @Override
                public void onCancel() {
                    //do nothing
                }

                @Override
                public void onEnd() {
                    number--;
                    if (number == 0) {
                        tvCountDown.setText("GO");
                        LAnimationUtil.Companion.play(tvCountDown, Techniques.Flash, new CallbackAnimation() {
                            @Override
                            public void onCancel() {
                                //do nothing
                            }

                            @Override
                            public void onEnd() {
                                tvCountDown.setVisibility(GONE);
                                //tvCountDown = null;
//                                Log.d(logTag, "number == 0 -> STOP");
                                if (callback != null) {
                                    callback.onEnd();
                                }
                            }

                            @Override
                            public void onRepeat() {
                                //do nothing
                            }

                            @Override
                            public void onStart() {
                                //do nothing
                            }
                        });
                        return;
                    }
                    doPerSec();
                }

                @Override
                public void onRepeat() {
                    //do nothing
                }

                @Override
                public void onStart() {
                    //do nothing
                    if (callback != null) {
                        callback.onTick();
                    }
                }
            });
        }
    }
}

File: core/src/main/java/com/views/textview/countdown/LCountDownView.kt
package com.views.textview.countdown

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import com.R
import com.core.utilities.LAnimationUtil
import com.daimajia.androidanimations.library.Techniques
import com.interfaces.CallbackAnimation
import kotlinx.android.synthetic.main.view_l_count_down.view.*

class LCountDownView : RelativeLayout {

    interface Callback {
        fun onTick()
        fun onEnd()
    }

    private val logTag = javaClass.simpleName
    private var number = 0
    private var callback: Callback? = null

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init()
    }

    private fun init() {
        inflate(context, R.layout.view_l_count_down, this)
    }

    fun start(number: Int) {
        this.number = number
        doPerSec()
    }

    fun setShowOrHide(isShow: Boolean) {
        tvCountDown.visibility = if (isShow) {
            VISIBLE
        } else {
            INVISIBLE
        }
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    private fun doPerSec() {
        tvCountDown.text = number.toString()
        LAnimationUtil.playDuration(
                view = tvCountDown,
                techniques = Techniques.FlipInX,
                duration = 1000,
                callbackAnimation = object : CallbackAnimation {
                    override fun onCancel() {
                        //do nothing
                    }

                    @SuppressLint("SetTextI18n")
                    override fun onEnd() {
                        number--
                        if (number == 0) {
                            tvCountDown.text = "GO"
                            LAnimationUtil.play(
                                    view = tvCountDown,
                                    techniques = Techniques.Flash,
                                    callbackAnimation = object : CallbackAnimation {
                                        override fun onCancel() {
                                            //do nothing
                                        }

                                        override fun onEnd() {
                                            tvCountDown?.visibility = GONE
                                            callback?.onEnd()
                                        }

                                        override fun onRepeat() {
                                            //do nothing
                                        }

                                        override fun onStart() {
                                            //do nothing
                                        }
                                    })
                            return
                        }
                        doPerSec()
                    }

                    override fun onRepeat() {
                        //do nothing
                    }

                    override fun onStart() {
                        callback?.onTick()
                    }
                })
    }
}
